From 2d22b551c1873811ecf84fc880b2931aa22d3c00 Mon Sep 17 00:00:00 2001
From: qiulaidongfeng <2645477756@qq.com>
Date: Tue, 23 Sep 2025 19:39:24 +0800
Subject: [PATCH] cmd/go: disable cgo by default if CC unset and DefaultCC doesn't exist

CL 621995 disrupted the behavior
introduced by CL 450739.

CL 703555 attempted to fix
but was incorrect.

The correct default value is
First use buildcfg.DefaultCGO_ENABLED.
If not set, use the behavior of go/build.defaultContext.
Final check if CC unset and Default CC doesn't exist,
disabled by default.

Fixes #75340

Change-Id: I3ea6b2b8b279521c0cc6c119520e9ae55c69fd59
---

diff --git a/src/cmd/go/internal/cfg/cfg.go b/src/cmd/go/internal/cfg/cfg.go
index a4edd85..21d9b73 100644
--- a/src/cmd/go/internal/cfg/cfg.go
+++ b/src/cmd/go/internal/cfg/cfg.go
@@ -147,33 +147,33 @@
 	} else if buildcfg.DefaultCGO_ENABLED == "0" {
 	} else if runtime.GOARCH == ctxt.GOARCH && runtime.GOOS == ctxt.GOOS {
 		defaultCgoEnabled = platform.CgoSupported(ctxt.GOOS, ctxt.GOARCH)
-		// Use built-in default cgo setting for GOOS/GOARCH.
-		// Note that ctxt.GOOS/GOARCH are derived from the preference list
-		// (1) environment, (2) go/env file, (3) runtime constants,
-		// while go/build.Default.GOOS/GOARCH are derived from the preference list
-		// (1) environment, (2) runtime constants.
-		//
-		// We know ctxt.GOOS/GOARCH == runtime.GOOS/GOARCH;
-		// no matter how that happened, go/build.Default will make the
-		// same decision (either the environment variables are set explicitly
-		// to match the runtime constants, or else they are unset, in which
-		// case go/build falls back to the runtime constants), so
-		// go/build.Default.GOOS/GOARCH == runtime.GOOS/GOARCH.
-		// So ctxt.CgoEnabled (== go/build.Default.CgoEnabled) is correct
-		// as is and can be left unmodified.
-		//
-		// All that said, starting in Go 1.20 we layer one more rule
-		// on top of the go/build decision: if CC is unset and
-		// the default C compiler we'd look for is not in the PATH,
-		// we automatically default cgo to off.
-		// This makes go builds work automatically on systems
-		// without a C compiler installed.
-		if ctxt.CgoEnabled {
-			if os.Getenv("CC") == "" {
-				cc := DefaultCC(ctxt.GOOS, ctxt.GOARCH)
-				if _, err := pathcache.LookPath(cc); err != nil {
-					defaultCgoEnabled = false
-				}
+	}
+	// Use built-in default cgo setting for GOOS/GOARCH.
+	// Note that ctxt.GOOS/GOARCH are derived from the preference list
+	// (1) environment, (2) go/env file, (3) runtime constants,
+	// while go/build.Default.GOOS/GOARCH are derived from the preference list
+	// (1) environment, (2) runtime constants.
+	//
+	// We know ctxt.GOOS/GOARCH == runtime.GOOS/GOARCH;
+	// no matter how that happened, go/build.Default will make the
+	// same decision (either the environment variables are set explicitly
+	// to match the runtime constants, or else they are unset, in which
+	// case go/build falls back to the runtime constants), so
+	// go/build.Default.GOOS/GOARCH == runtime.GOOS/GOARCH.
+	// So ctxt.CgoEnabled (== go/build.Default.CgoEnabled) is correct
+	// as is and can be left unmodified.
+	//
+	// All that said, starting in Go 1.20 we layer one more rule
+	// on top of the go/build decision: if CC is unset and
+	// the default C compiler we'd look for is not in the PATH,
+	// we automatically default cgo to off.
+	// This makes go builds work automatically on systems
+	// without a C compiler installed.
+	if ctxt.CgoEnabled {
+		if os.Getenv("CC") == "" {
+			cc := DefaultCC(ctxt.GOOS, ctxt.GOARCH)
+			if _, err := pathcache.LookPath(cc); err != nil {
+				defaultCgoEnabled = false
 			}
 		}
 	}
